### PR TITLE
Dev jbcoe descending expression matcher

### PR DESCRIPTION
**Pull request by jbcoe**

### CODE REVIEW

File: src/matcher.rs [hunk 1]
```diff
@@ -2,6 +2,7 @@
 // this module is called matcher
 
 #![allow(clippy::module_name_repetitions)]
+#![allow(unused_macros)]
```
Comment (philipcraig) [inline src/matcher.rs]: please don't merge this

File: src/matcher.rs [hunk 2]
```diff
@@ -2,6 +2,7 @@
 // this module is called matcher
 
 #![allow(clippy::module_name_repetitions)]
```
Comment (philipcraig) [inline src/matcher.rs]: I want to maintain no clippy warnings. To remove this one, I either need to rename the module from `matcher` to something else, or change the `struct`s to not have `Matcher` as a prefix or suffix. Which would you prefer?

Note that we had no issue when the module was named `ast_matcher`.

Comment (jbcoe) [reply src/matcher.rs]: I'm unconvinced by the value in this warning. XMatcher is the right name for a thing that matches an X, if clippy wants me to rename the (correctly named) file I'm happy silencing clippy (on this issue).

Comment (jbcoe) [reply src/matcher.rs]: Unconvinced by the value in this clippy warning.

File: src/matcher.rs (hunk 1 above)
Comment (jbcoe) [reply src/matcher.rs]: agreed. fixed by later PR.

**Comment by philipcraig** [comment]
Why the force pushes? :(

**Review by philipcraig — APPROVED** [review]
LGTM

**Pull request merged**